File: configs/velite/gallery/equipment/armor.ts
import { s } from "velite";

export const gallery_equipment_armor_data = {
    name: "GalleryEquipmentArmorData",
    pattern: "gallery/_data/equipment/armor/**/*.yml",
    schema: s
        .object({
            basic: s.object({
                name: s.string().max(32),
                quality: s.enum(["D", "C", "B", "A", "S"]),
                job: s.enum(["cike", "fashi", "mushi", "sheshou", "zhanshi"]),
                image: s.string().optional(), // 对应 base.image
            }),
            // 基础数值（耐久 / 护甲）
            value: s
                .object({
                    durable: s.number().positive().optional(), // max-item-damage
                    armor: s.number().nonnegative().optional(),
                    "armor-toughness": s.number().nonnegative().optional(),
                })
                .optional(),
            // 加成与特殊效果（全部可选）
            effect: s
                .object({
                    "max-health": s.number().optional(), // 最大生命加成
                    defense: s.number().optional(), // 防御减伤
                    "max-mana": s.number().optional(), // 法力加成
                    "max-stamina": s.number().optional(), // 耐力加成
                    "parry-rating": s.number().optional(), // 招架几率
                    "movement-speed": s.number().optional(), // 移速加成
                    "dodge-rating": s.number().optional(), // 闪避率
                })
                .optional(),
            gem: s
                .object({
                    count: s.number().int().nonnegative().optional(), // weapon-card
                    volume: s.number().int().nonnegative().optional(), // weapon-swordvolume
                })
                .optional(),
        })
        .transform((data, { meta }) => {
            // 路径解析：gallery/_data/equipment/armor/CIKE/A/BOOTS.yml
            // → job=CIKE, quality=A, part=BOOTS
            const segments = meta.path.replace(/\.yml$/, "").split("\\");
            const part = segments.at(-1) ?? "UNKNOWN";
            const qualityFromPath = segments.at(-2) ?? data.basic.quality;
            const jobFromPath = segments.at(-3)?.toLowerCase() ?? data.basic.job;

            return {
                id: `${jobFromPath}-${qualityFromPath}-${part}`.toLowerCase(),
                ...data,
            };
        }),
};

File: scripts/gallery-generator/gallery_generator/generators/armor.py
"""护甲 gallery 生成器。"""

from __future__ import annotations

import logging
import re
from pathlib import Path
from typing import Any

from ..utils import write_yaml_with_merge

logger = logging.getLogger(__name__)

name = "armor"
help = "从 armor.yml 生成护甲 gallery 数据"

PART_MAP: dict[str, str] = {
    "KUI": "HELMET", "XIONG": "CHESTPLATE", "KU": "LEGGINGS", "XUE": "BOOTS",
}

KEY_PATTERN = re.compile(r"^([A-Z]+)_(D|C|B|A|S)_([A-Z]+)$")
GRADIENT_RE = re.compile(r"<gradient:[^>]+>")


def strip_gradient(name: str) -> str:
    return GRADIENT_RE.sub("", name).strip()


def transform_item(key: str, base: dict[str, Any]) -> tuple[str, str, str, dict[str, Any]] | None:
    m = KEY_PATTERN.match(key)
    if not m:
        logger.warning("跳过不符合命名规则的键: %s", key)
        return None
    job_upper, quality, part_code = m.groups()
    part = PART_MAP.get(part_code)
    if part is None:
        logger.warning("未知部位代码 %s，跳过键: %s", part_code, key)
        return None
    basic: dict[str, Any] = {
        "name": strip_gradient(str(base.get("name", ""))),
        "quality": quality,
        "job": job_upper.lower(),
    }
    if "image" in base:
        basic["image"] = base["image"]
    value: dict[str, Any] = {}
    if "max-item-damage" in base:
        value["durable"] = base["max-item-damage"]
    if "armor" in base:
        value["armor"] = base["armor"]
    if "armor-toughness" in base:
        value["armor-toughness"] = base["armor-toughness"]
    effect: dict[str, Any] = {}
    for k in ("max-health", "defense", "max-mana", "max-stamina", "parry-rating", "movement-speed", "dodge-rating"):
        if k in base:
            effect[k] = base[k]
    gem: dict[str, Any] = {}
    if "weapon-card" in base:
        gem["count"] = base["weapon-card"]
    if "weapon-swordvolume" in base:
        gem["volume"] = base["weapon-swordvolume"]
    result: dict[str, Any] = {"basic": basic}
    if value:
        result["value"] = value
    if effect:
        result["effect"] = effect
    if gem:
        result["gem"] = gem
    return job_upper, quality, part, result


def run(src: Path, dst: Path) -> int:
    """从 armor.yml 生成 armor gallery 文件。"""
    import yaml

    if not src.is_file():
        raise FileNotFoundError(f"源文件不存在: {src}")
    with src.open("r", encoding="utf-8") as f:
        data: Any = yaml.safe_load(f)
    if not isinstance(data, dict):
        raise ValueError(f"YAML 根节点不是字典: {src}")
    dst.mkdir(parents=True, exist_ok=True)
    count = 0
    for key, item in data.items():
        if not isinstance(item, dict) or "base" not in item:
            continue
        base = item["base"]
        if not isinstance(base, dict):
            continue
        transformed = transform_item(key, base)
        if transformed is None:
            continue
        job, quality, part, content = transformed
        out_path = dst / job.lower() / quality.lower() / f"{part.lower()}.yml"
        out_path.parent.mkdir(parents=True, exist_ok=True)
        write_yaml_with_merge(out_path, content)
        logger.info("已写入: %s", out_path.name)
        count += 1
    return count